### PR TITLE
加入分隔符

### PR DESCRIPTION
**Pull request by LeiZhang-Hunter**
单行测试配置
```
pipelines:
  - name: local
    sources:
      - type: file
        name: demo
        lineDelimiter:
          type: auto
        paths:
          - /home/zhanglei/*.log
        fields:
          topic: "loggie"
    interceptors:
      - type: iconv
    sink:
      type: dev
      printEvents: true
      codec:
        pretty: true
```

### CODE REVIEW

File: pkg/source/file/process/line.go [hunk 1]
```diff
@@ -33,7 +33,8 @@ func (lp *LineProcessor) Process(processorChain file.ProcessChain, ctx *file.Job
 	read := int64(len(readBuffer))
 	processed := int64(0)
 	for processed < read {
-		index := int64(bytes.IndexByte(readBuffer[processed:], '\n'))
+		split := job.GetSplit()
+		index := int64(bytes.IndexAny(readBuffer[processed:], split))
```
Comment (JeffPhoneix) [inline pkg/source/file/process/line.go:37]: https://pkg.go.dev/bytes#IndexAny

File: pkg/source/file/split.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+var globalSplit Split
+
+func (split *Split) Init() {
+	if split.split == nil {
+		split.split = make(map[string]string)
+	}
+}
+
+func (split *Split) AddSplit(pipelineName string, sourceName string, value string) error {
```
Comment (JeffPhoneix) [inline pkg/source/file/split.go:20]: 这些函数可能都是多协程中调用的，实现需要保证多协程安全。

File: pkg/source/file/split.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	if len(value) == 0 {
+		return errors.New(fmt.Sprintf("%s:%s split is empty", pipelineName, sourceName))
+	}
+	key := fmt.Sprintf("%s:%s", pipelineName, sourceName)
+	split.split[key] = value
+	return nil
+}
+
+func (split *Split) GetSplit(pipelineName string, sourceName string) string {
+	key := fmt.Sprintf("%s:%s", pipelineName, sourceName)
```
Comment (JeffPhoneix) [inline pkg/source/file/split.go:30]: 数据面上尽量不用fmt.Sprintf来拼接字符串。如strings.Builder?

**Comment by LeiZhang-Hunter** [comment]
https://blog.csdn.net/qq_32783703/article/details/125413088

**Comment by LeiZhang-Hunter** [comment]
this is test report!
https://blog.csdn.net/qq_32783703/article/details/125413088

File: pkg/interceptor/iconv/interceptor.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	}
+	return invoker.Invoke(invocation)
+}
+
+func (i *Interceptor) process(e api.Event) error {
+	if i.config.Charset == "utf-8" {
+		return nil
+	}
+
+	codec, ok := util.AllEncodings[i.config.Charset]
```
Comment (mmaxiaolei) [inline pkg/interceptor/iconv/interceptor.go:88]: It can be initialized in the `Init()` phase

File: pkg/interceptor/iconv/interceptor.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+func (i *Interceptor) Intercept(invoker source.Invoker, invocation source.Invocation) api.Result {
+	e := invocation.Event
+	if err := i.process(e); err != nil {
+		log.Error("encode event %s error: %v", e.String(), err)
+	}
+	return invoker.Invoke(invocation)
+}
+
+func (i *Interceptor) process(e api.Event) error {
+	if i.config.Charset == "utf-8" {
```
Comment (mmaxiaolei) [inline pkg/interceptor/iconv/interceptor.go:84]: It can be initialized in the Init() phase

File: pkg/interceptor/iconv/interceptor.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+	}
+
+	codec, ok := util.AllEncodings[i.config.Charset]
+
+	if !ok {
+		log.Warn("unknown Charset('%v')", i.config.Charset)
+		return errors.New(fmt.Sprintf("unknown Charset('%v')", i.config.Charset))
+	}
+
+	bytes, err := codec.NewDecoder().Bytes(e.Body())
```
Comment (mmaxiaolei) [inline pkg/interceptor/iconv/interceptor.go:95]: `Decoder` can be initialized in the Init() phase

File: pkg/source/file/end_line.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+	if end.endLineMap == nil {
+		end.endLineMap = make(map[string]*LineEndingsValue)
+	}
+}
+
+func (end *LineEndings) AddLineEnd(pipelineName string, sourceName string, lineEndValue *LineEndValue) error {
+	end.mutex.Lock()
+	defer end.mutex.Unlock()
+	lineType, ok := lineTerminators[lineEndValue.LineType]
+	if ok == false && lineEndValue.LineType != Custom {
```
Comment (mmaxiaolei) [inline pkg/source/file/end_line.go:92]: `ok == false` can simple to `!ok`

File: pkg/source/file/end_line.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+func (end *LineEndings) GetLineEnd(pipelineName string, sourceName string) []byte {
+	end.mutex.RLock()
+	defer end.mutex.RUnlock()
+	var key strings.Builder
+	key.WriteString(pipelineName)
+	key.WriteString(":")
+	key.WriteString(sourceName)
+	value, ok := end.endLineMap[key.String()]
+	if ok == false {
```
Comment (mmaxiaolei) [inline pkg/source/file/end_line.go:132]: `ok == false` can simple to `!ok`

File: pkg/source/file/job.go [hunk 9]
```diff
@@ -292,11 +292,19 @@ func (j *Job) File() *os.File {
...
+	return globalLineEnd.GetEncodeLineEnd(j.task.pipelineName, j.task.sourceName)
+}
+
+func (j *Job) GetLineEnd() []byte {
+	return globalLineEnd.GetLineEnd(j.task.pipelineName, j.task.sourceName)
+}
+
 func (j *Job) ProductEvent(endOffset int64, collectTime time.Time, body []byte) {
-	nextOffset := endOffset + 1
+	nextOffset := endOffset + int64(len(j.GetEncodeLineEnd()))
```
Comment (mmaxiaolei) [inline pkg/source/file/job.go:304]: `int64(len(j.GetEncodeLineEnd()))` can be initialized in the `newJob()` to achieve lock free. `Productevent()` is the core data link. We need to be very careful about performance

File: pkg/source/file/multilines.go [hunk 10]
```diff
@@ -117,8 +117,9 @@ func (mh *MultiHolder) appendContent(content []byte, state State) {
 		mh.state = state
 	} else {
 		// has content, add '\n'
-		mh.content = append(mh.content, '\n')
-		mh.currentSize++
+		split := globalLineEnd.GetLineEnd(mh.state.PipelineName, mh.state.SourceName)
```
Comment (mmaxiaolei) [inline pkg/source/file/multilines.go:120]: `split` can be initialized in the `newMultiHolder`

File: pkg/source/file/multilines.go [hunk 11]
```diff
@@ -141,7 +142,7 @@ func (mh *MultiHolder) flush() {
 	if mh.currentSize <= 0 {
 		return
 	}
-	mh.state.ContentBytes = mh.currentSize + 1
+	mh.state.ContentBytes = mh.currentSize + int64(len(globalLineEnd.GetEncodeLineEnd(mh.state.PipelineName, mh.state.SourceName)))
```
Comment (mmaxiaolei) [inline pkg/source/file/multilines.go:145]: `int64(len(globalLineEnd.GetEncodeLineEnd(mh.state.PipelineName, mh.state.SourceName)))` can be initialized in the `newMultiHolder`

File: pkg/source/file/source.go [hunk 12]
```diff
@@ -116,6 +117,7 @@ func (s *Source) Init(context api.Context) error {
 		SourceName:   s.name,
 		Level:        IsolationLevel(s.config.Isolation),
 	}
+	globalLineEnd.AddLineEnd(s.pipelineName, s.name, &s.config.LineEnd)
```
Comment (mmaxiaolei) [inline pkg/source/file/source.go:120]: Called in the `component.Init()` phase, because the config may not be initialized at this time

**Pull request merged**